README.md
# YAIOM
Factorio mod, Yet Another Infinite Ore Mechanic

Having issues finding reliable ore spots farther and farther away? Want some sustainable but non-trivial ore sources?

Look no further.

## News

- Nerfed patches so they now use double fracking fluid. Sorry. **BUT** I added a mod option, Equivalent Exchange, to re-enable the old behaviour, so no worries!
- Renamed the ore to Orichalcum, as the name fits better. I'll miss Ferricupric ;-;
- Added interfacing with Angel's Refining, more forthcoming...

## Hydraulic fracturing

When you generate a map with this mod on, you may notice that there's blue/green/pink/... ores on the map generation preview, but once you're within the game, you can't find them.

![](https://i.imgur.com/3dhc8rB.png)

Well, that's because they're generated deep underground, and you'll need to extract them if you want to reach them. And scan them first for their location so you can even see them.

Researching **Hydraulic fracturing** (fracking) will unlock a *Deep soil scanner* that scans the floor for gravimetric fluctuations deep underground, in the same radius as a basegame radar's active coverage (7x7 chunks). The miners for the ores are unlocked with further research (once you research the means to process any deep ore).

![](https://i.imgur.com/c2vENpj.png)

*Note: Ore quantities are exaggerated on this image for display purposes.*

Place one down and let it run, as they take a while. Each scanner takes at most around an hour in real time to finish scanning to its fullest width (2 minutes per chunk), and then deconstruct itself once done.

Deep soil scanners won't scan chunks that have already been scanned (and revealed any deep ores), so spamming them around the terrain is the most efficient way to uncover the hidden ore.

### Satellite coverage

If you're tired of having to place and move scanners around, there's the **Orbital deep soil scanning** research which unlocks the means to use satellites to also scan the soil over time, getting faster the more satellites you have.

But to make this happen you'll need to place a *Satellite beacon* on the surface you wish to scan and wait, as each passive scan takes a whole Factorio day (~7 minutes), **but can reveal several chunks at once** depending on how many satellites you have in orbit. Just, don't place more than one beacon at a time. Trust me on this one.

## Orichalcum ore

The main mythical ore you find deep underground, being a mixture of hydro-carbonates and traces of both weak ferric and cupric acids. It shows on the minimap as a light blue (or light green, if you have Angel's Refining installed), for ease of visibility. **In the basegame, it can produce copper, iron, oil side-products and eventually coal.** On modded savegames, it will spawn further ores, usually biased for starting and/or massively needed resources.

Requires **light oil** to extract it from the floor, where high yields will require less oil per 1lu of orichalcum sludge (at the mininum yield, 50%, you get 1 sludge per 2 light oil used).

An early game setup for the sludge may look like this:

![early setup](https://i.imgur.com/vy0ENu9.png)

While a late game setup, requiring more research, may look like this:

![late setup](https://i.imgur.com/Qo4IvKz.jpg)

*Note*: The earlier setup may still be of use even after unlocking the advanced ore refining, seeing that you get a higher return of petroleum gas than with further refining.

## Yet to come

- Better graphics. I know, tinting stuff blue is my signature by now but it looks groddy.
- Adding dedicated support for big modpacks, such as adding a separate deep ore for tin-lead or integrating with Angel's ore processing more tightly.
  - Already started with Angel's refining, see [this chart]().
- RSO compatability. Will come with 2.0.0, I promise, but it'll require a rewrite so may take a tad.

## Bugs

Report on [my github](https://github.com/dustine/factorio-yaiom) or on [the forum post for this mod](https://forums.factorio.com/56737). Don't forget to include your savegame, the log file(s), a zip folder with all of your currently using mods, and to describe what you were doing before the bug happened.

## Credits
- [Katherine of Skies](https://www.youtube.com/channel/UCTIV3KbAvaGEyNjoMoNaGtQ)'s community for the mod idea and support
- Various members on the Factorio Discord and IRC for handling my rambling about
- Arch666Angel, for the placeholder graphic for the beacon

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 1.4.2
Date: 05.08.2018
  Bugfixes:
    - Fixed modded productivity modules crashing the mod
---------------------------------------------------------------------------------------------------
Version: 1.4.1
Date: 02.02.2018
  Balancing:
    - Swapped the need for catalyst on Irradiated Orichalcum recipes (Angel's Refining integration) based on how more valuable Crotinnium and Jivolite are
  Bugfixes:
    - Fixed changelog having 1.4 stuff listed as 1.3, twice
---------------------------------------------------------------------------------------------------
Version: 1.4.0
Date: 02.02.2018
  Major features:
    - Added integration for Angel Refining (the one that adds the new ores)
    - Renamed the ore to Orichalcum, which was an obvious name for it in retrospect...
  Balancing:
    - Doubled light oil usage, so <100% spots will now take more fracking fluid than the sludge they give back (down to 50%, so 2lu per 1lu of Orichalcum).
    - Added setting to revert to old fracking fluid usage
    - Patches are slightly less rich now (to incentive exploration just a tad)
  Bugfixes:
    - Removed RSO support... for now
---------------------------------------------------------------------------------------------------
Version: 1.3.0
Date: 27.01.2018
  Minor features:
    - Added RSO support
    - Added ru locale (thank you konikovsky!)
---------------------------------------------------------------------------------------------------
Version: 1.2.3
Date: 24.01.2018
  Bugfixes:
    - Advanced processing only exported 50 ore per crafting (game limitation, changed recipes but kept throughput as a workaround).
    - Reworked how variables are indexed so I'm 100% sure desyncs can't be a thing now. For sure.
    - Beacons took x60 times longer to scan than intended.
  Balancing:
    - Increased uranium usage slightly, because of the item stack size bugfix. Sorry!
---------------------------------------------------------------------------------------------------
Version: 1.2.2
Date: 22.01.2018
  Balancing:
    - Increased ferricupric mining speed by 2.5x (faster mining speed and power on miners).
    - Hydraulic miners now can have up to 4 modules now.
    - Advanced processing uses 1000 ore (x4) per enriched uranium, but takes twice as long now.
---------------------------------------------------------------------------------------------------
Version: 1.2.1
Date: 19.01.2018
  Bugfixes:
    - Possible desync over invalid surface reference.
---------------------------------------------------------------------------------------------------
Version: 1.2.0
Date: 19.01.2018
  Minor features:
    - Already existing savegames (without the mod) will retroactively contain deep ore (hidden or not).
  Bugfixes:
    - Hidden ores now spawn by default on modded surfaces (autoplace.default_enabled is now true).
---------------------------------------------------------------------------------------------------
Version: 1.1.0
Date: 18.01.2018
  Major features:
    - Research to allow satellites to scan for deep ore passively (using a beacon to activate this feature).
---------------------------------------------------------------------------------------------------
Version: 1.0.0
Date: 17.01.2018
  Major features:
    - Hidden ores that require deep soil scanners to reveal them, and time.
    - Hydraulic cracking, allows to mine hidden ores with high pressure liquid injection.
    - Ferricupric ore, special hidden ore that can be refined into copper and iron, and coal with further research.
  Minor features:
    - Setting to disable satellite hidden ore discovery.

data-updates.lua
local presets = data.raw["map-gen-presets"].default

presets["rich-resources"].basic_settings.autoplace_controls["yaiom-orichalcum"] = { richness = "very-good"}
presets["rail-world"].basic_settings.autoplace_controls["yaiom-orichalcum"] = {
  frequency = "very-low",
  size = "high"
}

local Data = require "stdlib.data.data"
local Item = require "stdlib.data.item"
local Entity = require "stdlib.data.entity"
local Recipe = require "stdlib.data.recipe"
local Technology = require "stdlib.data.technology"

local function make_recipe_production_enabled(recipe)
  for name, module in pairs(data.raw.module) do
    if name:find("productivity") and module.limitation then
      table.insert(module.limitation, recipe)
    end
  end
end

local function finish_sorting_recipe(name, results, subgroup, icon_product, icon_result, icon_result_behind)
  local icon = {{
    icon = "__angelsrefining__/graphics/icons/sort-icon.png",
  }}
  local recipe = Recipe(name)
    :change_category("ore-sorting")
    :subgroup_order(subgroup)
    :set_fields({
      results = {}
    })
    :replace_icon(icon, 32)
  if icon_product then
    table.insert(icon, {
      icon = icon_product,
      shift = {-8,8},
      scale = 0.5,
    })
  end
  if icon_result_behind then
    table.insert(icon, {
      icon = icon_result_behind,
      shift = {8,4},
      scale = 0.5,
    })
  end
  if icon_result then
    table.insert(icon, {
      icon = icon_result,
      shift = {8,8},
      scale = 0.5,
    })
  end
  for _, result in pairs(results) do
    table.insert(recipe.results, result)
  end
  return recipe
end

if angelsmods and angelsmods.refining then
  -- pick a teal green so to avoid confusing with the blue ores of angel's
  Entity("yaiom-orichalcum", "resource")
    :set_fields({
      map_color = util.color "00FF7F"
    })

  Data.extend_array({{
    type = "item-subgroup",
    name = "yaiom-sorting",
    group = "resource-refining",
    order = "j-z[yaiom]-b[orichalcum]"
  }})

  --##########################################################################--
  --                                  TIER 0                                  --
  --##########################################################################--

  local filtering = Recipe("yaiom-orichalcum-filtration")
    :change_category("filtering")
    :add_ingredient({type = "fluid", name = "water-purified", amount = 50})
    :add_ingredient({type = "item", name = "filter-coal", amount = 1})
    :set_fields({
      subgroup = "filtering"
    })
    :replace_icon({{
      icon = "__angelsrefining__/graphics/icons/filter-coal.png",
    },{
      icon = "__yaiom__/graphics/orichalcum/sludge.png",
      shift = {10,8},
      scale = 0.6,
    }}, 32)
  table.insert(filtering.results, {type="item", name="filter-frame", amount=1})

  Recipe("yaiom-orichalcum-cleaning")
    :change_category("liquifying")
    :subgroup_order("liquifying", "z[yaiom]-b[orichalcum]-0-b[degassing]")
    :set_fields({
      localised_name = {"recipe-name.yaiom-orichalcum-degassing"},
      ingredients = {
        {type = "fluid", name = "yaiom-orichalcum-sludge", amount = 30}
      },
      results = {
        {type = "fluid", name = "petroleum-gas", amount = 20},
        {type = "item", name = "slag", amount = 10}
      },
    })
    :replace_icon("__yaiom__/graphics/orichalcum/recipe-degas.png", 32)

  local technology_0 = Technology("yaiom-orichalcum")
    :remove_pack("production-science-pack")
    :remove_prereq("advanced-material-processing-2")
    :add_prereq("slag-processing-1")
    :add_effect("yaiom-orichalcum-filtration")
  technology_0.effects[2], technology_0.effects[3] = technology_0.effects[3], technology_0.effects[2]

  --##########################################################################--
  --                                  TIER 1                                  --
  --##########################################################################--

  -- degassing.icon = "__yaiom__/graphics/orichalcum/dust.png"
  local dust = Item("yaiom-orichalcum")
    -- TODO: right category
    :subgroup_order(nil, "z[yaiom]-b[orichalcum]-1-a[dust]")
    :set_fields({
      localised_name = {"item-name.yaiom-orichalcum-dust"}
    })

  local process = Recipe("yaiom-orichalcum-processing")
    :change_category("water-treatment")
    :subgroup_order("water-cleaning", "z[yaiom]-b[orichalcum]-1-a[processing]")
    :remove_ingredient("water")
    :replace_ingredient("yaiom-orichalcum-fluid", "yaiom-orichalcum-sludge")
    :set_fields({
      localised_name = {"recipe-name.yaiom-orichalcum-dust"},
      energy_required = 3,
    })
    :replace_icon({{
      icon = "__yaiom__/graphics/orichalcum/sludge.png",
      shift = {0,-4},
      scale = 0.75
    },{
      icon = "__base__/graphics/icons/fluid/light-oil.png",
      shift = {-11,11},
      scale = 0.25
    },{
      icon = dust.icon,
      shift = {0,12},
      scale = 0.20
    },{
      icon = "__base__/graphics/icons/coal.png",
      dark_background_icon = "__base__/graphics/icons/coal-dark-background.png",
      shift = {12,12},
      scale = 0.20
    }}, 32)
  process.main_product = "yaiom-orichalcum"
  table.insert(process.results, {type = "fluid", name = "water-purified", amount = 50})

  finish_sorting_recipe("yaiom-orichalcum-dust-sorting", {{"angels-ore1-crushed", 1}, {"angels-ore3-crushed", 1}, {"slag", 1}}, "yaiom-sorting", dust.icon)
    :set_fields({
      localised_name = {"recipe-name.yaiom-sorting-2", dust.localised_name, {"entity-name.angels-ore1"}, {"entity-name.angels-ore3"}}
    })

  Recipe("yaiom-orichalcum-iron")
    :remove_unlock("yaiom-advanced-orichalcum-1")

  Recipe("yaiom-orichalcum-copper")
    :remove_unlock("yaiom-advanced-orichalcum-1")

  Technology("yaiom-advanced-orichalcum-1")
    :remove_pack("high-tech-science-pack")
    :remove_prereq("nuclear-fuel-reprocessing")
    :add_prereq("advanced-material-processing-2")
    :add_prereq("water-treatment-2")
    :add_effect("yaiom-orichalcum-dust-sorting")
    :set_fields({count = 1000})

  --##########################################################################--
  --                                  TIER 2                                  --
  --##########################################################################--

  local ore1_icon = Data.get_icon_from("item", "angels-ore1-crushed")
  local ore2_icon = Data.get_icon_from("item", "angels-ore2-crushed")
  local ore3_icon = Data.get_icon_from("item", "angels-ore3-crushed")
  local ore4_icon = Data.get_icon_from("item", "angels-ore4-crushed")

  local irradiated = Item("yaiom-irradiated-orichalcum")
    :remove_flag("hidden")
    :subgroup_order()

  local irradiated_name = {"item-name.yaiom-irradiated-orichalcum"}

  finish_sorting_recipe("yaiom-orichalcum-ore-sorting-n-1", {{"angels-ore1-crushed", 4}, {"angels-ore4-crushed", 1}}, "yaiom-sorting", irradiated.icon, ore1_icon, ore4_icon)
    :set_fields({
      localised_name = {"recipe-name.yaiom-sorting-2", irradiated_name, {"entity-name.angels-ore1"}, {"entity-name.angels-ore4"}}
    })

  finish_sorting_recipe("yaiom-orichalcum-ore-sorting-n-2", {{"angels-ore3-crushed", 4}, {"angels-ore2-crushed", 1}}, "yaiom-sorting", irradiated.icon, ore3_icon, ore2_icon)
    :set_fields({
      localised_name = {"recipe-name.yaiom-sorting-2", irradiated_name, {"entity-name.angels-ore3"}, {"entity-name.angels-ore2"}}
    })

  finish_sorting_recipe("yaiom-orichalcum-ore-sorting-b-1", {{"angels-ore2-crushed", 2}, {"angels-ore4-crushed", 2}}, "yaiom-sorting", irradiated.icon, ore2_icon, ore4_icon)
    :add_ingredient("catalysator-brown")
    :set_fields({
      localised_name = {"recipe-name.yaiom-sorting-2", irradiated_name, {"entity-name.angels-ore2"}, {"entity-name.angels-ore4"}}
    })

  Technology("yaiom-advanced-orichalcum-2")
    :set_fields({enabled = true,})

  make_recipe_production_enabled("yaiom-orichalcum-cleaning")
  make_recipe_production_enabled("yaiom-orichalcum-processing")
else
  for technology in pairs(data.raw.technology) do
    Recipe("fill-yaiom-orichalcum-sludge-barrel")
      :remove_unlock()
    Recipe("empty-yaiom-orichalcum-sludge-barrel")
      :remove_unlock()
  end

  make_recipe_production_enabled("yaiom-orichalcum-cleaning")
  make_recipe_production_enabled("yaiom-orichalcum-processing")
end